engine/src/main/java/nl/inl/blacklab/search/results/HitsFromQuery.java

```diff
@@ -337,8 +337,8 @@ protected void ensureResultsRead(int number) {
                             hitQueryContext.getCapturedGroups(groups);
                             hitsResultsContext.capturedGroups.put(hit, groups);
                         }
-                        hitsResultsContext.results.add(hit);
-                        if (maxHitsToProcess >= 0 && hitsResultsContext.results.size() >= maxHitsToProcess) {
+                        hitsResultsContext.hitsArrays.add(hit);
```

Comment:
There is a lock around the call to `add` that isn't needed on this path. We may want a version of `add` that doesn't lock.